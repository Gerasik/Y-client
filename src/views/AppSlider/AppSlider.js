export default class AppMain {
  constructor(move, lock) {
    this.name = 'slider';
    this.element = null;
    this.lock = lock;
    this.move = move;
  }

  render() {
    if (!document.querySelector(`.${this.name}`)) {
      const slider = document.createElement('section');
      slider.classList.add(this.name);
      document.body.appendChild(slider);
    }
    this.addPagination();
    this.addSlide();

    document.querySelector(`.${this.name}`).addEventListener('mousedown', this.lock, false);
    document.querySelector(`.${this.name}`).addEventListener('touchstart', this.lock, false);
    document.querySelector(`.${this.name}`).addEventListener('mouseup', this.move, false);
    document.querySelector(`.${this.name}`).addEventListener('touchend', this.move, false);
  }

  addSlide() {
    this.element = document.querySelector(`.${this.name}`);
    const slide = document.createElement('div');
    slide.classList.add(`${this.name}__slide`);
    this.element.appendChild(slide);
  }

  addPagination() {
    const paginationBar = document.createElement('ul');
    const buttonPrev = document.createElement('li');
    const buttonCurrPage = document.createElement('li');
    const buttonNextPage = document.createElement('li');
    const buttonNext = document.createElement('li');

    paginationBar.classList.add('pagination');
    buttonPrev.classList.add('pagination__prev');
    buttonCurrPage.classList.add('pagination__curr');
    buttonNextPage.classList.add('pagination__next-page');
    buttonNext.classList.add('pagination__next');

    buttonPrev.innerHTML = '<button><</button>';
    buttonCurrPage.innerHTML = '<button>1</button>';
    buttonNextPage.innerHTML = '<button>2</button>';
    buttonNext.innerHTML = '<button>></button>';

    paginationBar.addEventListener('click', this.move, false);

    document.body.appendChild(paginationBar);
    paginationBar.appendChild(buttonPrev);
    paginationBar.appendChild(buttonCurrPage);
    paginationBar.appendChild(buttonNextPage);
    paginationBar.appendChild(buttonNext);

    window.addEventListener('resize', () => {
      let clipsCount;
      if (window.outerWidth > 1300) {
        clipsCount = 4;
      } else if (window.outerWidth <= 1300 && window.outerWidth > 768) {
        clipsCount = 3;
      } else if (window.outerWidth <= 768 && window.outerWidth > 425) {
        clipsCount = 2;
      } else if (window.outerWidth <= 425) {
        clipsCount = 1;
      }
      const dataList = [];
      while (this.element.childElementCount) {
        while (this.element.firstChild.childElementCount) {
          dataList.push(this.element.firstChild.removeChild(this.element.firstChild.firstChild));
        }
        this.element.firstChild.remove();
      }
      this.addSlide();
      dataList.forEach((elem) => {
        if (this.element.lastChild.childElementCount >= clipsCount) {
          this.addSlide();
        }
        this.element.lastChild.appendChild(elem);
      });
    });
  }
}
