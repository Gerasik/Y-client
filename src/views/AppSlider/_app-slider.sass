.slider
  width: 100%
  overflow: hidden
  display: flex
  &__slide
    display: flex
    min-width: 100%
    height: 100%
    margin: 25px 0
    justify-content: space-around
    align-items: stretch
    transition: 0.8s all cubic-bezier(0.74, 0.02, 0.17, 1.01)
.pagination
  display: flex
  justify-content: center
  li
    list-style: none
    button
      height: 25px
      border-radius: 50%
      border: none
      padding: 5px
      margin: 5px
      width: 25px
      background: #dc7a15
      color: white
      font-weight: bold