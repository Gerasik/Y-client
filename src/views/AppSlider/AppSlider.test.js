import AppSlider from './AppSlider';

describe('AppSlider.prototype.render', () => {
  it('Should be an instance of Function', () => {
    expect(AppSlider.prototype.render).toBeInstanceOf(Function);
  });

  it('Should be render correctly', () => {
    const context = {
      lock: function a() {
        return 1;
      },
      move: function b() {
        return 2;
      },
      addPagination: function c() {
        return 3;
      },
      addSlide: function d() {
        return 4;
      },
    };
    AppSlider.prototype.render.call(context);
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('Should be render slide', () => {
    const context = {
      element: document.body,
    };
    AppSlider.prototype.addSlide.call(context);
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('Should be render slider pagination', () => {
    const context = {
      element: document.body,
    };
    AppSlider.prototype.addPagination.call(context);
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
