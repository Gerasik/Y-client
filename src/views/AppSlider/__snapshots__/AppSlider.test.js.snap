// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppSlider.prototype.render Should be render correctly 1`] = `"<section class=\\"undefined\\"></section>"`;

exports[`AppSlider.prototype.render Should be render slide 1`] = `"<section class=\\"undefined\\"><div class=\\"undefined__slide\\"></div></section>"`;

exports[`AppSlider.prototype.render Should be render slider pagination 1`] = `"<section class=\\"undefined\\"><div class=\\"undefined__slide\\"></div></section><ul class=\\"pagination\\"><li class=\\"pagination__prev\\"><button>&lt;</button></li><li class=\\"pagination__curr\\"><button>1</button></li><li class=\\"pagination__next-page\\"><button>2</button></li><li class=\\"pagination__next\\"><button>&gt;</button></li></ul>"`;
